libflux/src/semantic/types.rs

```diff
@@ -240,82 +821,114 @@ pub struct Function {
                 k: k.clone(),
                 v: v.clone(),
-            });
-        }
-        req.sort_unstable_by(|a, b| a.k.cmp(&b.k));
+            })
+            .collect::<Vec<_>>();
 
-        let mut opt: Vec<Property> = Vec::new();
-        for (k, v) in &self.opt {
-            opt.push(Property {
+        let optional = self
+            .opt
+            .iter()
+            .collect::<BTreeMap<_, _>>()
+            .iter()
+            .map(|(&k, &v)| Property {
                 k: String::from("?") + &k,
                 v: v.clone(),
-            });
-        }
-        opt.sort_unstable_by(|a, b| a.k.cmp(&b.k));
-
-        let mut args: Vec<Property> = Vec::new();
-        if let Some(pipe) = &self.pipe {
-            if pipe.k == "<-" {
-                args.push(pipe.clone());
-            } else {
-                args.push(Property {
-                    k: String::from("<-") + &pipe.k,
-                    v: pipe.v.clone(),
-                });
+            })
+            .collect::<Vec<_>>();
+
+        let pipe = match &self.pipe {
+            Some(pipe) => {
+                if pipe.k == "<-" {
+                    vec![pipe.clone()]
+                } else {
+                    vec![Property {
+                        k: String::from("<-") + &pipe.k,
+                        v: pipe.v.clone(),
+                    }]
+                }
             }
-        }
+            None => vec![],
+        };
+
+        let values = pipe
+            .iter()
+            .chain(required.iter().chain(optional.iter()))
+            .collect();
 
-        args.append(&mut req);
-        args.append(&mut opt);
         write!(
             f,
             "({}) -> {}",
             DisplayList {
-                values: &args,
-                delim: ", "
+                values: values,
+                delim: ", ",
             },
             self.retn
         )
     }
 }
 
-// BoundTvar represents a constrained type variable.
-// Used solely for displaying the generic constraints of a polytype.
-#[derive(Debug)]
-struct BoundTvar {
-    tv: Tvar,
-    kind: Kind,
+impl Substitutable for HashMap<String, MonoType> {
+    fn apply(self, sub: &Subst) -> Self {
+        let mut args = self;
+        for (_, v) in &mut args {
+            *v = v.clone().apply(sub);
+        }
+        args
+    }
 }
 
-impl fmt::Display for BoundTvar {
-    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
-        write!(f, "{}:{}", self.tv, self.kind)
+impl Substitutable for Function {
+    fn apply(self, sub: &Subst) -> Self {
+        Function {
+            req: self.req.apply(sub),
+            opt: self.opt.apply(sub),
+            pipe: match self.pipe {
+                None => None,
+                Some(p) => Some(p.apply(sub)),
+            },
+            retn: self.retn.apply(sub),
+        }
+    }
+}
+
+impl Unifiable for Function {
+    fn unify(self, _: Self, _: Const) -> Result {
+        unimplemented!();
```

Comment:
those `unimplemented!` are left for future implementation in another PR?

Reply:
Yes correct.